scripts/hid_configurator/led_stream.py

```diff
@@ -65,5 +75,187 @@ def send_continuous_led_stream(dev, recipient, DEVICE_CONFIG, led_id, freq, subs
+
+        send_led_effects.duration_increment = duration_inc_variable
+        if in_buf_speaker > in_buf_dev:
+            send_led_effects.duration_increment += DURATION_INCREMENT_CONST_FACTOR
+        if in_buf_speaker < in_buf_dev:
+            send_led_effects.duration_increment -= DURATION_INCREMENT_CONST_FACTOR
+
+        send_led_effects.duration_increment *= send_led_effects.BASE_DURATION
+
+
+def music_callback(in_data, frame_count, time_info, status):
+    if status == pyaudio.paOutputUnderflow:
+        print("Underflow occurred. Please lower the stream frequency.")
+        return (None, pyaudio.paAbort)
+
+    if music_callback.send_error_event.is_set():
+        print("Device communication problem occurred")
+        return (None, pyaudio.paAbort)
+
+    data = music_callback.file.readframes(frame_count)
+    music_callback.queue.put(data)
+
+    return (data, pyaudio.paContinue)
+
+
+def send_music_led_stream(dev, recipient, DEVICE_CONFIG, led_id,
+                          freq, filename, substep_cnt = 10):
+    if not validate_params(DEVICE_CONFIG, freq, led_id):
+        return
+
+    success, (free, ready) = fetch_free_steps_buffer_info(dev, recipient, led_id)
+    if not success:
+        print("Device communication problem occurred")
+        return
+
+    if not ready:
+        print("LEDs are not ready")
+        return
+
+    try:
+        wf = wave.open(filename, 'rb')
+    except (wave.Error, FileNotFoundError) as e:
+        print(e)
+        return
+
+    music_callback.file = wf
+
+    send_led_effects.max_free = free
+    send_led_effects.dev = dev
+    send_led_effects.recipient = recipient
+    send_led_effects.led_id = led_id
+    send_led_effects.BASE_DURATION = 1 / freq
+
+    send_led_effects.sent_cnt = 0
+    send_led_effects.reminder = 0
+    send_led_effects.duration_increment = 0
+
+    evt = threading.Event()
+    q = queue.Queue()
+    music_callback.queue = q
+    music_callback.send_error_event = evt
+
+    p = pyaudio.PyAudio()
+    stream = p.open(format=p.get_format_from_width(wf.getsampwidth()),
+                    channels=wf.getnchannels(),
+                    rate=wf.getframerate(),
+                    output=True,
+                    input=False,
+                    stream_callback=music_callback,
+                    frames_per_buffer=wf.getframerate() // freq,
+                    start = False)
+
+    send_led_effects.out_latency = stream.get_output_latency()
+
+    try:
+        print("Music LED stream started. Press Ctrl+C to interrupt.")
+        threading.Thread(target=send_led_effects, args=(evt, q)).start()
+        send_led_effects.start_time = time.time()
+        stream.start_stream()
+        while stream.is_active():
+            time.sleep(0.1)
+
+    except KeyboardInterrupt:
+        print("Music LED stream interrupted by user")
+        pass
+
+    print("Music LED stream ended")
+    stream.stop_stream()
+    stream.close()
+    p.terminate()
+    wf.close()
```

Comment:
All resources must be contained within 'with' statements. This probably also will require you to create a function to reduce too many nested levels.

Reply:
This code will be moved to class (according to @ppryga suggestion)